File: Features/encumbrance-avatar/contracts/src/test/java/net/corda/samples/avatar/contracts/ContractTests.java
package net.corda.samples.avatar.contracts;

import com.google.common.collect.ImmutableList;
import net.corda.core.identity.CordaX500Name;
import net.corda.samples.avatar.states.Avatar;
import net.corda.samples.avatar.states.Expiry;
import net.corda.testing.core.TestIdentity;
import net.corda.testing.node.MockServices;
import org.junit.Test;

import java.time.Duration;
import java.time.Instant;
import java.time.temporal.ChronoUnit;

import static net.corda.testing.node.NodeTestUtils.ledger;

public class ContractTests {

    static private final MockServices ledgerServices = new MockServices();
    static private final TestIdentity seller = new TestIdentity(new CordaX500Name("MegaCorp", "London", "GB"));
    static private final TestIdentity buyer = new TestIdentity(new CordaX500Name("MiniCorp", "London", "GB"));


    //Both the encumbrance and the encumbered state must be added to the transaction
    @Test
    public void thereMustBeTwoOutputs() {
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.output(AvatarContract.AVATAR_CONTRACT_ID, new Avatar(seller.getParty(), "1"));
                tx.command(ImmutableList.of(seller.getPublicKey()), new AvatarContract.Commands.Create());
                tx.fails();
                return null;
            });
            return null;
        }));
    }


    @Test
    public void encumbranceIndexMustBeSpecified() {
        //not specifying the encumbrance index fails the contract
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.output(AvatarContract.AVATAR_CONTRACT_ID, new Avatar(seller.getParty(), "1"));
                tx.output(ExpiryContract.EXPIRY_CONTRACT_ID,
                        new Expiry(Instant.now().plus(2, ChronoUnit.MINUTES), "1", seller.getParty()));
                tx.command(ImmutableList.of(seller.getPublicKey()), new AvatarContract.Commands.Create());
                tx.timeWindow(Instant.now(), Duration.ofMinutes(1));
                tx.failsWith("Avatar needs to be encumbered");

                return null;
            });
            return null;
        }));

        //specifying the encumbrance index while adding output states passes the contract
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.output(AvatarContract.AVATAR_CONTRACT_ID, 1, new Avatar(seller.getParty(), "1"));
                tx.output(ExpiryContract.EXPIRY_CONTRACT_ID, 0,
                        new Expiry(Instant.now().plus(2, ChronoUnit.MINUTES), "1", seller.getParty()));
                tx.command(ImmutableList.of(seller.getPublicKey()), new AvatarContract.Commands.Create());
                tx.timeWindow(Instant.now(), Duration.ofMinutes(1));
                tx.verifies();
                return null;
            });
            return null;
        }));
    }

    //Specifying time window is mandatory. This is checked in the encumbrance Expiry state.
    @Test
    public void specifyTimeWindow() {
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                //this fails as time window is not specified
                tx.output(AvatarContract.AVATAR_CONTRACT_ID, 1, new Avatar(seller.getParty(), "1"));
                tx.output(ExpiryContract.EXPIRY_CONTRACT_ID, 0, new Expiry(Instant.now().plus(2, ChronoUnit.MINUTES), "1", seller.getParty()));
                tx.command(ImmutableList.of(seller.getPublicKey()), new AvatarContract.Commands.Create());
                tx.fails();

                //this will pass once we specify time window
                tx.timeWindow(Instant.now(), Duration.ofMinutes(1));
                return tx.verifies();
            });
            return null;
        }));
    }

    //For selling, the Expiry of avatar must be greater than the time window
    @Test
    public void avatarIsRejectedIfItIsExpired() {
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.output(AvatarContract.AVATAR_CONTRACT_ID, 1, new Avatar(seller.getParty(), "1"));
                tx.output(ExpiryContract.EXPIRY_CONTRACT_ID, 0, new Expiry(Instant.now().plus(2, ChronoUnit.MINUTES), "1", seller.getParty()));
                tx.command(ImmutableList.of(seller.getPublicKey()), new AvatarContract.Commands.Create());
                tx.timeWindow(Instant.now(), Duration.ofMinutes(3));
                tx.fails();
                return null;
            });
            return null;
        }));
    }

    //For selling, the Expiry of avatar must be greater than the time window
    @Test
    public void expirationDateShouldBeAfterTheTimeWindow() {
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.output(AvatarContract.AVATAR_CONTRACT_ID, 1, new Avatar(seller.getParty(), "1"));
                tx.output(ExpiryContract.EXPIRY_CONTRACT_ID, 0, new Expiry(Instant.now().plus(3, ChronoUnit.MINUTES), "1", seller.getParty()));
                tx.command(ImmutableList.of(seller.getPublicKey()), new AvatarContract.Commands.Create());
                tx.timeWindow(Instant.now(), Duration.ofMinutes(2));
                tx.verifies();
                return null;
            });
            return null;
        }));
    }

    //test transaction which has encumbered states as inputs
    @Test
    public void transferAvatar() {
        ledger(ledgerServices, ledger -> {
            ledger.unverifiedTransaction(tx -> {
                tx.output(AvatarContract.AVATAR_CONTRACT_ID, "avatarLabel", 1,
                        new Avatar(seller.getParty(), "1"));
                tx.output(ExpiryContract.EXPIRY_CONTRACT_ID, "expiryLabel", 0,
                        new Expiry(Instant.now().plus(3, ChronoUnit.MINUTES), "1", seller.getParty()));
                return null;
            });
            ledger.transaction(tx -> {
                tx.input("avatarLabel");
                tx.input("expiryLabel");
                tx.output(AvatarContract.AVATAR_CONTRACT_ID, "avatarLabel2", 1,
                        new Avatar(buyer.getParty(), "1"));
                tx.output(ExpiryContract.EXPIRY_CONTRACT_ID, "expiryLabel2", 0,
                        new Expiry(Instant.now().plus(3, ChronoUnit.MINUTES), "1", buyer.getParty()));
                tx.command(ImmutableList.of(seller.getPublicKey(), buyer.getPublicKey()), new AvatarContract.Commands.Transfer());
                tx.command(ImmutableList.of(seller.getPublicKey(), buyer.getPublicKey()), new ExpiryContract.Commands.Pass());
                tx.timeWindow(Instant.now(), Duration.ofMinutes(2));
                tx.verifies();
                return null;
            });
            return null;
        });
    }


    @Test
    public void avatarCannotBeSpentWithoutExpiry() {
        ledger(ledgerServices, ledger -> {
            ledger.unverifiedTransaction(tx -> {
                tx.output(AvatarContract.AVATAR_CONTRACT_ID, "avatarLabel", 1,
                        new Avatar(seller.getParty(), "1"));
                tx.output(ExpiryContract.EXPIRY_CONTRACT_ID, "expiryLabel", 0,
                        new Expiry(Instant.now().plus(3, ChronoUnit.MINUTES), "1", seller.getParty()));
                return null;
            });
            ledger.transaction(tx -> {
                tx.input("avatarLabel");
//                tx.input(expiryLabel);
                tx.output(AvatarContract.AVATAR_CONTRACT_ID, "avatarLabel2", 1,
                        new Avatar(buyer.getParty(), "1"));
                tx.output(ExpiryContract.EXPIRY_CONTRACT_ID, "expiryLabel2", 0,
                        new Expiry(Instant.now().plus(3, ChronoUnit.MINUTES), "1", buyer.getParty()));
                tx.command(ImmutableList.of(seller.getPublicKey(), buyer.getPublicKey()), new AvatarContract.Commands.Transfer());
                tx.command(ImmutableList.of(seller.getPublicKey(), buyer.getPublicKey()), new ExpiryContract.Commands.Pass());
                tx.timeWindow(Instant.now(), Duration.ofMinutes(2));
                tx.fails();
                return null;
            });
            return null;
        });
    }
}

File: kotlin-samples/Features/encumbrance-avatar/contracts/src/test/kotlin/net/corda/samples/avatar/contracts/ContractTests.kt
package net.corda.samples.avatar.contracts

import net.corda.core.identity.CordaX500Name
import net.corda.samples.avatar.contracts.ExpiryContract.Commands.Pass
import net.corda.samples.avatar.states.Avatar
import net.corda.samples.avatar.states.Expiry
import net.corda.testing.core.TestIdentity
import net.corda.testing.node.MockServices
import net.corda.testing.node.ledger
import org.junit.Test
import java.time.Duration
import java.time.Instant
import java.time.temporal.ChronoUnit

class ContractTests {
    private val ledgerServices: MockServices = MockServices(listOf("net.corda.samples.avatar"))
    var seller = TestIdentity(CordaX500Name("Alice", "TestLand", "US"))
    var buyer = TestIdentity(CordaX500Name("Alice", "TestLand", "US"))

    @Test
    fun thereMustBeTwoOutputs() {
        ledgerServices.ledger {
            // Should fail bid price is equal to previous highest bid
            transaction {
                //failing transaction
                output(AvatarContract.AVATAR_CONTRACT_ID, Avatar(seller.party, "1"))
                command(listOf(seller.publicKey), AvatarContract.Commands.Create())
                fails()
            }
        }
    }

    @Test
    fun encumbranceIndexMustBeSpecified() {
        //not specifying the encumbrance index fails the contract
        ledgerServices.ledger {
            transaction {
                output(AvatarContract.AVATAR_CONTRACT_ID, Avatar(seller.party, "1"))
                output(
                    ExpiryContract.EXPIRY_CONTRACT_ID,
                    Expiry(Instant.now().plus(2, ChronoUnit.MINUTES), "1", seller.party)
                )
                command(listOf(seller.publicKey), AvatarContract.Commands.Create())
                timeWindow(Instant.now(), Duration.ofMinutes(1))
                failsWith("Avatar needs to be encumbered")
            }
        }
        //specifying the encumbrance index while adding output states passes the contract
        ledgerServices.ledger {
            transaction {
                output(AvatarContract.AVATAR_CONTRACT_ID, 1, Avatar(seller.party, "1"))
                output(
                    ExpiryContract.EXPIRY_CONTRACT_ID,
                    0,
                    Expiry(Instant.now().plus(2, ChronoUnit.MINUTES), "1", seller.party)
                )
                command(listOf(seller.publicKey), AvatarContract.Commands.Create())
                timeWindow(Instant.now(), Duration.ofMinutes(1))
                verifies()
            }
        }
    }

    //Specifying time window is mandatory. This is checked in the encumbrance Expiry state.
    @Test
    fun specifyTimeWindow() {
        ledgerServices.ledger {
            transaction {
                //this fails as time window is not specified
                output(AvatarContract.AVATAR_CONTRACT_ID, 1, Avatar(seller.party, "1"))
                output(
                    ExpiryContract.EXPIRY_CONTRACT_ID,
                    0,
                    Expiry(Instant.now().plus(2, ChronoUnit.MINUTES), "1", seller.party)
                )
                command(listOf(seller.publicKey), AvatarContract.Commands.Create())
                fails()
            }
            transaction {
                output(AvatarContract.AVATAR_CONTRACT_ID, 1, Avatar(seller.party, "1"))
                output(
                    ExpiryContract.EXPIRY_CONTRACT_ID,
                    0,
                    Expiry(Instant.now().plus(2, ChronoUnit.MINUTES), "1", seller.party)
                )
                command(listOf(seller.publicKey), AvatarContract.Commands.Create())
                //this will pass once we specify time window
                timeWindow(Instant.now(), Duration.ofMinutes(1))
                verifies()
            }
        }
    }

    //For selling, the Expiry of avatar must be greater than the time window
    @Test
    fun avatarIsRejectedIfItIsExpired() {
        ledgerServices.ledger {
            transaction {
                output(AvatarContract.AVATAR_CONTRACT_ID, Avatar(seller.party, "1"))
                output(
                    ExpiryContract.EXPIRY_CONTRACT_ID,
                    Expiry(Instant.now().plus(2, ChronoUnit.MINUTES), "1", seller.party)
                )
                command(listOf(seller.publicKey), AvatarContract.Commands.Create())
                timeWindow(Instant.now(), Duration.ofMinutes(3))
                fails()
            }
        }
    }

    //For selling, the Expiry of avatar must be greater than the time window
    @Test
    fun expirationDateShouldBeAfterTheTimeWindow() {
        ledgerServices.ledger {
            transaction {
                output(AvatarContract.AVATAR_CONTRACT_ID, 1, Avatar(seller.party, "1"))
                output(
                    ExpiryContract.EXPIRY_CONTRACT_ID,
                    0,
                    Expiry(Instant.now().plus(3, ChronoUnit.MINUTES), "1", seller.party)
                )
                command(listOf(seller.publicKey), AvatarContract.Commands.Create())
                timeWindow(Instant.now(), Duration.ofMinutes(2))
                verifies()
            }
        }
    }

    //test transaction which has encumbered states as inputs
    @Test
    fun transferAvatar() {
        ledgerServices.ledger {
            unverifiedTransaction {
                output(AvatarContract.AVATAR_CONTRACT_ID, "avatarLabel", 1, Avatar(seller.party, "1"))
                output(
                    ExpiryContract.EXPIRY_CONTRACT_ID,
                    "expiryLabel",
                    0,
                    Expiry(Instant.now().plus(3, ChronoUnit.MINUTES), "1", seller.party)
                )
            }
            transaction {
                input("avatarLabel")
                input("expiryLabel")
                output(AvatarContract.AVATAR_CONTRACT_ID, "avatarLabel2", 1, Avatar(buyer.party, "1"))
                output(
                    ExpiryContract.EXPIRY_CONTRACT_ID,
                    "expiryLabel2",
                    0,
                    Expiry(Instant.now().plus(3, ChronoUnit.MINUTES), "1", buyer.party)
                )
                command(listOf(seller.publicKey, buyer.publicKey), AvatarContract.Commands.Transfer())
                command(listOf(seller.publicKey, buyer.publicKey), Pass())
                timeWindow(Instant.now(), Duration.ofMinutes(2))
                verifies()
            }
        }
    }


    @Test
    fun avatarCannotBeSpentWithoutExpiry() {
        ledgerServices.ledger {
            unverifiedTransaction {
                output(AvatarContract.AVATAR_CONTRACT_ID, "avatarLabel", 1, Avatar(seller.party, "1"))
                output(
                    ExpiryContract.EXPIRY_CONTRACT_ID,
                    "expiryLabel",
                    0,
                    Expiry(Instant.now().plus(3, ChronoUnit.MINUTES), "1", seller.party)
                )
            }
            transaction {
                input("avatarLabel")
                output(AvatarContract.AVATAR_CONTRACT_ID, "avatarLabel2", 1, Avatar(buyer.party, "1"))
                output(
                    ExpiryContract.EXPIRY_CONTRACT_ID,
                    "expiryLabel2",
                    0,
                    Expiry(Instant.now().plus(3, ChronoUnit.MINUTES), "1", buyer.party)
                )
                command(listOf(seller.publicKey, buyer.publicKey), AvatarContract.Commands.Transfer())
                command(listOf(seller.publicKey, buyer.publicKey), Pass())
                timeWindow(Instant.now(), Duration.ofMinutes(2))
                fails()
            }
        }
    }

}